### DISCUSSION TITLE
8088 vs 8086 prefetch queue

### DISCUSSION THREAD

**Discussion by Torinde** [dbalsom/marty#108] [Q&A] [answered]
https://github.com/adriancable/8086tiny/issues/5#issuecomment-640315873
> There are differences between 8086 and 8088 visible to self-modifying code due to the different size of the prefetch queue (IIRC, 6 bytes on 8086 but 4 on 8088). This does not depend on bus or memory timing, and is AFAIK the only reliable way for a program to distinguish 8086 from 8088.
Most simulators don't bother simulating the prefetch queue at all, let alone accurately, in which case code attempting to determine the prefetch queue length will fail.

Does MartyPC emulate that?

**Comment by dbalsom** [comment, marked as answer]
Yes. CPU detection routines will properly identify MartyPC as an 8088 if they use the queue length detection method.

You can view the contents of the instruction queue from the CPU Status window.

**Reply by Torinde** [reply]
Good! 
As far as I saw it currently runs only as 8088. Is there an option to run it as 8086?

**Reply by dbalsom** [reply]
Not at the moment, but that should be coming fairly soon. I have to rewrite the 8088 core one last time, and during that rewrite I will look into preliminary 8086 support. I originally had a bunch of checks in the core to determine 8088 vs 8086 behavior, but every conditional within a tight CPU loop slows down the emulator quite a bit.

The execution unit of the 8088 and 8086 are mostly the same, it is the bus interface unit that differs. So I am thinking to use generics to compose an appropriate CPU based on swapping out the BIU. I will also need to add support for multiple CPUs without dragging down performance either.
